src/state_tracker/state.rs

```diff
@@ -66,77 +66,66 @@ where
+                self.state[last_index] = MapValue(S::from(label)); //TODO: looks similar!
             },
-            (Some(MapKey(Some(oldlabel))), String(label)) => {
-                if oldlabel.as_ref() >= label {
-                    return self.latch_err(Err(E::from(StructureError::UnsortedKeys)));
-                }
-                self.state.push(MapValue(S::from(label)));
+            (Some(MapKey(Some(oldlabel))), String(label)) if oldlabel.as_ref() >= label => {
+                self.state.pop();
+                return self.latch_err(Err(E::from(StructureError::UnsortedKeys)));
+            },
+            (Some(MapKey(Some(_oldlabel))), String(label)) => {
+                self.state[last_index] = MapValue(S::from(label));
             },
-            (Some(oldstate @ MapKey(_)), _tok) => {
-                self.state.push(oldstate);
+            (Some(_oldstate @ MapKey(_)), _tok) => {
+                self.state.pop();
                 return self.latch_err(Err(E::from(StructureError::invalid_state(
                     "Map keys must be strings",
                 ))));
             },
-            (Some(MapValue(label)), List) => {
-                self.state.push(MapKey(Some(label)));
-                if self.state.len() >= self.max_depth {
-                    return self.latch_err(Err(E::from(StructureError::NestingTooDeep)));
-                }
-                self.state.push(Seq);
-            },
-            (Some(MapValue(label)), Dict) => {
-                self.state.push(MapKey(Some(label)));
+            (Some(MapValue(label)), List) | (Some(MapValue(label)), Dict) => {
+                let dummy: &[u8] = &[1];
+                self.state[last_index] = MapKey(Some(std::mem::replace(label, dummy.into())));
                 if self.state.len() >= self.max_depth {
                     return self.latch_err(Err(E::from(StructureError::NestingTooDeep)));
                 }
-                self.state.push(MapKey(None));
+                self.state
+                    .push(if token == &List { Seq } else { MapKey(None) });
             },
-            (Some(oldstate @ MapValue(_)), End) => {
-                self.state.push(oldstate);
+            (Some(_oldstate @ MapValue(_)), End) => {
+                self.state.pop();
                 return self.latch_err(Err(E::from(StructureError::invalid_state(
                     "Missing map value",
                 ))));
             },
             (Some(MapValue(label)), _) => {
-                self.state.push(MapKey(Some(label)));
+                let dummy: &[u8] = &[1];
+                self.state[last_index] = MapKey(Some(std::mem::replace(label, dummy.into())));
             },
-            (oldstate, List) => {
-                if let Some(oldstate) = oldstate {
-                    self.state.push(oldstate);
+            (oldstate, List) | (oldstate, Dict) => {
+                if oldstate.is_none() && 0 < self.state.len() {
```

Comment:
The only way that oldstate would be None is if self.state is empty, so this will never evaluate to true.